File: programs/perp-margin-accounts/src/instructions/execute_withdraw.rs
use crate::errors::MarginError;
use crate::state::{MarginAccount, MarginVault};
use crate::util::fees::process_fees;
use crate::util::pnl::process_pnl_update;
use crate::util::validate::validate_balances;
use anchor_lang::prelude::*;
use anchor_spl::token::{self, Token, TokenAccount, Transfer};
use perp_amm::{program::PerpAmm, state::PoolState};

#[derive(Accounts)]
pub struct ExecuteWithdrawal<'info> {
    #[account(
        mut,
        seeds = [b"margin_account", margin_account.owner.as_ref()],
        bump = margin_account.bump,
        constraint = margin_account.pending_sol_withdrawal > 0 || margin_account.pending_usdc_withdrawal > 0 @ MarginError::NoPendingWithdrawal
    )]
    pub margin_account: Account<'info, MarginAccount>,

    #[account(
        seeds = [b"margin_vault"],
        bump = margin_vault.bump,
        constraint = authority.key() == margin_vault.authority @ MarginError::InvalidAuthority
    )]
    pub margin_vault: Account<'info, MarginVault>,

    #[account(
        mut,
        constraint = margin_sol_vault.key() == margin_vault.margin_sol_vault
    )]
    pub margin_sol_vault: Account<'info, TokenAccount>,

    #[account(
        mut,
        constraint = margin_usdc_vault.key() == margin_vault.margin_usdc_vault
    )]
    pub margin_usdc_vault: Account<'info, TokenAccount>,

    #[account(
        mut,
        constraint = user_sol_account.owner == margin_account.owner
    )]
    pub user_sol_account: Account<'info, TokenAccount>,

    #[account(
        mut,
        constraint = user_usdc_account.owner == margin_account.owner
    )]
    pub user_usdc_account: Account<'info, TokenAccount>,

    /// The liquidity pool's state account
    #[account(
        mut,
        seeds = [b"pool_state".as_ref()],
        bump
    )]
    pub pool_state: Account<'info, PoolState>,

    /// The liquidity pool's vault account that matches the token being withdrawn
    #[account(
        mut,
        constraint = pool_vault_account.key() == pool_state.sol_vault || pool_vault_account.key() == pool_state.usdc_vault
    )]
    pub pool_vault_account: Account<'info, TokenAccount>,

    /// CHECK: Validated in constraint against stored value in margin vault
    #[account(address = margin_vault.chainlink_program)]
    pub chainlink_program: AccountInfo<'info>,

    /// CHECK: Validated in constraint against stored value in margin vault
    #[account(address = margin_vault.chainlink_feed)]
    pub chainlink_feed: AccountInfo<'info>,

    #[account(
        constraint = authority.key() == margin_vault.authority @ MarginError::UnauthorizedExecution
    )]
    pub authority: Signer<'info>,

    pub token_program: Program<'info, Token>,
    pub liquidity_pool_program: Program<'info, PerpAmm>,
    pub system_program: Program<'info, System>,
}

// Split the function to reduce stack usage
pub fn handler(
    mut ctx: Context<ExecuteWithdrawal>,
    pnl_update: i64,
    locked_sol: u64,
    locked_usdc: u64,
    sol_fees_owed: u64,
    usdc_fees_owed: u64,
) -> Result<()> {
    // Process fees
    process_fees(
        &mut ctx.accounts.margin_account,
        &mut ctx.accounts.margin_vault,
        sol_fees_owed,
        usdc_fees_owed,
    )?;

    // Validate balances against locked amounts
    validate_balances(&ctx.accounts.margin_account, locked_sol, locked_usdc)?;

    // Process PnL updates if needed
    if pnl_update != 0 {
        process_pnl_update(&mut ctx, pnl_update)?;
    }

    // Process withdrawals
    process_withdrawals(&mut ctx)?;

    Ok(())
}

// Helper function to process withdrawals
fn process_withdrawals(ctx: &mut Context<ExecuteWithdrawal>) -> Result<()> {
    let margin_account = &mut ctx.accounts.margin_account;

    // Process SOL withdrawal if pending.
    if margin_account.pending_sol_withdrawal > 0 {
        let sol_amount = margin_account.pending_sol_withdrawal;
        margin_account.sol_balance = margin_account
            .sol_balance
            .checked_sub(sol_amount)
            .ok_or(MarginError::ArithmeticOverflow)?;

        let seeds = &[b"margin_vault".as_ref(), &[ctx.accounts.margin_vault.bump]];
        let signer = &[&seeds[..]];

        let cpi_accounts = Transfer {
            from: ctx.accounts.margin_sol_vault.to_account_info(),
            to: ctx.accounts.user_sol_account.to_account_info(),
            authority: ctx.accounts.margin_vault.to_account_info(),
        };
        let cpi_ctx = CpiContext::new_with_signer(
            ctx.accounts.token_program.to_account_info(),
            cpi_accounts,
            signer,
        );
        token::transfer(cpi_ctx, sol_amount)?;
    }

    // Process USDC withdrawal if pending.
    if margin_account.pending_usdc_withdrawal > 0 {
        let usdc_amount = margin_account.pending_usdc_withdrawal;
        margin_account.usdc_balance = margin_account
            .usdc_balance
            .checked_sub(usdc_amount)
            .ok_or(MarginError::ArithmeticOverflow)?;

        let seeds = &[b"margin_vault".as_ref(), &[ctx.accounts.margin_vault.bump]];
        let signer = &[&seeds[..]];

        let cpi_accounts = Transfer {
            from: ctx.accounts.margin_usdc_vault.to_account_info(),
            to: ctx.accounts.user_usdc_account.to_account_info(),
            authority: ctx.accounts.margin_vault.to_account_info(),
        };
        let cpi_ctx = CpiContext::new_with_signer(
            ctx.accounts.token_program.to_account_info(),
            cpi_accounts,
            signer,
        );
        token::transfer(cpi_ctx, usdc_amount)?;
    }

    // Clear pending withdrawals.
    margin_account.pending_sol_withdrawal = 0;
    margin_account.pending_usdc_withdrawal = 0;

    Ok(())
}

File: scripts/perp-margin-accounts/execute-withdraw.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { PerpMarginAccounts } from "../../target/types/perp_margin_accounts";
import { PerpAmm } from "../../target/types/perp_amm";
import { PublicKey, SystemProgram } from "@solana/web3.js";
import { TOKEN_PROGRAM_ID } from "@solana/spl-token";
import BN from "bn.js";
import * as dotenv from "dotenv";

dotenv.config();

// Chainlink program and feed constants
const CHAINLINK_PROGRAM_ID = new PublicKey(
  process.env.CHAINLINK_PROGRAM_ID ||
    "HEvSKofvBgfaexv23kMabbYqxasxU3mQ4ibBMEmJWHny"
);

const CHAINLINK_SOL_FEED = process.env.CHAINLINK_SOL_FEED
  ? new PublicKey(process.env.CHAINLINK_SOL_FEED)
  : new PublicKey("99B2bTijsU6f1GCT73HmdR7HCFFjGMBcPZY6jZ96ynrR");

async function main() {
  // Configure the client to use the specified cluster
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const marginProgram = anchor.workspace
    .PerpMarginAccounts as Program<PerpMarginAccounts>;

  const ammProgram = anchor.workspace.PerpAmm as Program<PerpAmm>;

  // Parse command line arguments
  // Arguments: owner_pubkey pnl_update locked_sol locked_usdc sol_fees_owed usdc_fees_owed
  const args = process.argv.slice(2);
  if (args.length < 6) {
    console.log(
      "Usage: ts-node execute-withdraw.ts <owner_public_key> <pnl_update> <locked_sol> <locked_usdc> <sol_fees_owed> <usdc_fees_owed>"
    );
    console.log(
      "Example: ts-node execute-withdraw.ts Aa1CeQKW8UJpXafJLQJXXCUqzxRwJKUJyeQst73xpTTK 0 0 0 0 0"
    );
    process.exit(1);
  }

  const ownerPublicKey = new PublicKey(args[0]);
  const pnlUpdate = new BN(args[1]);
  const lockedSol = new BN(args[2]);
  const lockedUsdc = new BN(args[3]);
  const solFeesOwed = new BN(args[4]);
  const usdcFeesOwed = new BN(args[5]);

  // Derive margin account PDA
  const [marginAccount] = PublicKey.findProgramAddressSync(
    [Buffer.from("margin_account"), ownerPublicKey.toBuffer()],
    marginProgram.programId
  );

  // Derive margin vault PDA
  const [marginVault] = PublicKey.findProgramAddressSync(
    [Buffer.from("margin_vault")],
    marginProgram.programId
  );

  // Derive pool state PDA
  const [poolState] = PublicKey.findProgramAddressSync(
    [Buffer.from("pool_state")],
    ammProgram.programId
  );

  // Fetch margin vault details to get token accounts
  const marginVaultData = await marginProgram.account.marginVault.fetch(
    marginVault
  );
  const marginSolVault = marginVaultData.marginSolVault;
  const marginUsdcVault = marginVaultData.marginUsdcVault;

  // Fetch pool state to get vault addresses
  const poolStateData = await ammProgram.account.poolState.fetch(poolState);
  const solVault = poolStateData.solVault;
  const usdcVault = poolStateData.usdcVault;

  // Fetch margin account to determine which vault account to use
  const marginAccountData = await marginProgram.account.marginAccount.fetch(
    marginAccount
  );

  // Determine which vault to use based on which token has a pending withdrawal
  let poolVaultAccount;
  if (marginAccountData.pendingSolWithdrawal.gt(new BN(0))) {
    poolVaultAccount = solVault;
  } else {
    poolVaultAccount = usdcVault;
  }

  try {
    // Get the user token accounts
    const userSolAccount = await anchor.utils.token.associatedAddress({
      mint: new PublicKey("So11111111111111111111111111111111111111112"),
      owner: ownerPublicKey,
    });

    // Get USDC mint from margin vault
    const usdcMint = poolStateData.usdcMint;
    const userUsdcAccount = await anchor.utils.token.associatedAddress({
      mint: usdcMint,
      owner: ownerPublicKey,
    });

    console.log(
      `Executing withdrawal for account: ${marginAccount.toString()}`
    );
    console.log(`PnL update: ${pnlUpdate.toString()}`);
    console.log(`Locked SOL: ${lockedSol.toString()}`);
    console.log(`Locked USDC: ${lockedUsdc.toString()}`);
    console.log(`SOL fees: ${solFeesOwed.toString()}`);
    console.log(`USDC fees: ${usdcFeesOwed.toString()}`);

    // Call the execute withdrawal instruction
    const tx = await marginProgram.methods
      .executeWithdrawal(
        pnlUpdate,
        lockedSol,
        lockedUsdc,
        solFeesOwed,
        usdcFeesOwed
      )
      .accountsStrict({
        marginAccount: marginAccount,
        marginVault: marginVault,
        marginSolVault: marginSolVault,
        marginUsdcVault: marginUsdcVault,
        userSolAccount: userSolAccount,
        userUsdcAccount: userUsdcAccount,
        poolState: poolState,
        poolVaultAccount: poolVaultAccount,
        chainlinkProgram: CHAINLINK_PROGRAM_ID,
        chainlinkFeed: CHAINLINK_SOL_FEED,
        authority: provider.wallet.publicKey,
        tokenProgram: TOKEN_PROGRAM_ID,
        liquidityPoolProgram: ammProgram.programId,
        systemProgram: SystemProgram.programId,
      })
      .signers([provider.wallet.payer])
      .rpc();

    console.log(`Transaction successful: ${tx}`);
    console.log(`Executed withdrawal for account: ${marginAccount.toString()}`);
  } catch (error) {
    console.error(`Error executing withdrawal:`, error);
  }
}

main().catch((error) => {
  console.error("Script failed:", error);
  process.exit(1);
});
